2258 tau profile

hiker opened the pull request:
Adds support for profiling using TAU. Fixes #2258

hiker commented:
Ready for review.

arporter commented on doc/user_guide/profiling.rst:
```diff
@@ -145,9 +149,10 @@ module. The functions that need to be implemented are described in
-exceptions are the template and simple_timing libraries,
-which are stand alone. The profiling example in
+with the corresponding 3rd party profiling tool, or use a compiler
+wrapper provided by the tool which will provided the required additional
```
s/provided/provide/

hiker replied:
Done.

arporter reviewed (changes requested):
Thanks very much Joerg, almost perfect.
Just a couple of tweaks required.
I was able to get gocean/eg5 working and have pprof report some results so all good. (I couldn't be bothered with attempting to get Java working for the GUI.)
There are no changes to PSyclone code so I won't run the integration tests.
The only Issue that mentions TAU (apart from 2258) is the one on potential PSyclone projects. For bonus points would you mind removing TAU from that?

hiker commented on lib/profiling/tau/README.md:
```diff
@@ -0,0 +1,100 @@
+
+## Dependencies
+
+TAU must be installed at link time, though not when compiling the
```
Done.

hiker commented on lib/profiling/tau/README.md:
```diff
@@ -0,0 +1,100 @@
+**Note:**
+
+- The ``<PATH-TO-PSYCLONE>`` differs depending on whether the wrapper
+  library is compiled in a clone of PSyclone repository or in a PSyclone
```
Done.

hiker commented on lib/profiling/tau/README.md:
````diff
@@ -0,0 +1,100 @@
+If you use the ``tau_f90.sh`` compiler wrapper, nothing else is required,
+otherwise you need to add the required libraries from TAU at link time.
+
+```shell
````
Done.

hiker commented on lib/profiling/tau/tau_psy.f90:
```diff
@@ -0,0 +1,137 @@
+  end subroutine profile_PSyDataStart
+
+  ! ---------------------------------------------------------------------------
+  !> Disable profiling
```
Good point. I had another look at the API, and actually fixed that. It now works as expected.

hiker commented on examples/gocean/eg5/profile/Makefile:
```diff
@@ -65,7 +65,7 @@ run: compile
 
```
Done.

hiker commented:
I've updated another README, and removed tau from the list of libraries to be compiled by default (since we are only supposed to build profile wrapper fully included in PSyclone, sorry, I missed that).

hiker commented:
I forgot to mention:
> For bonus points would you mind removing TAU from that?

I've removed tau and some other things that are already done from the project ticket - I want my bonus points :)

arporter commented:
> I've removed tau and some other things that are already done from the project ticket - I want my bonus points :)

I hereby award @hiker 10 (ten) bonus points for outstanding services to PSyclone development.

arporter reviewed (changes requested):
Very nearly there. The one markdown file you updated (thanks) seems to be missing some text. Given that lib/profiling/Makefile is affected, I decided I probably had better run the integration tests so I've fired those off.

hiker commented on lib/profiling/README.md:
```diff
@@ -164,15 +164,40 @@ it should already exist. For example:
-and ``nvidia`` profiling-library interfaces since they require external
+**Note** that compilation currently does not include ``dl_timer``, ``drhook``,
+``tau``, and ``nvidia`` profiling-library interfaces since they require external
 libraries to be available. For the same reason, the ``all`` target
```
As far as I can see that is already broken on trunk. I've fixed this ... and actually, now that I have rewritten and pushed this, I realised it was actually correct - you just have to continue reading the ```make all``  command in the next line and then next paragraph. Anyway, I think it's easier to read now, since obviously this was easy to misunderstand.

arporter replied:
Oh yes. Thanks for improving it :-)

hiker commented:
Ready for next review, assuming that CI agrees.
